### DISCUSSION TITLE
Redis max client/connections hit

### DISCUSSION THREAD

**Discussion by jdgriffith** [taskforcesh/bullmq#1355] [General]
Your library is amazing and I appreciate all the updates and effort put into this over the years. 🙌🏻

Our current setup requires us to run 1 million + jobs in a distributed fashion. We currently use docker-compose and AWS ECS (Elastic Container Service) for these setups. For the current discussion, know that I am starting 5 docker containers in a docker-compose stack, with each container having one queue instantiated and 4 BullMQ workers with a concurrency of 4. 

```
        const worker = new Worker(
            queueName,
            async (job: Job) => this.processJob(job),
            {
                connection: this.getRedis(),
                concurrency: 4,
            },
        );
```

After kicking off the map-reduce jobs, around 1 million jobs are enqueued and the workers begin processing jobs. In a matter of minutes, the 10k max connections setting in Redis is hit and starts erroring out. I would not expect to hit that threshold with this current setup, although there is a high number of jobs. 

Is BullMQ creating a new connection for every job to be processed or something close to that? Are there new workers being created beyond the 4 I created in my code logic? Are jobs needing their own connections (possibly close to 1:1) to communicate status back to Redis?  

I understand the workers need to use blocking connections but I would expect the number of blocking connections to maybe be 1:1 with the concurrency setting I have (currently set at 4).

**Comment by manast** [comment]
The concurrency factor does not increase the number of connections, nor does the number of connections increase as the jobs are being processed. This sound most likely as you are creating new connections inside the processor function or something like that. If you can provide a minimal test case that shows the number of connections increasing as the jobs are processed you are welcome to create an issue and we can look deeper into it.

**Reply by jdgriffith** [reply]
Interesting. Let me take a closer look at our code and see if I can reproduce or determine the issue.

**Comment by jdgriffith** [comment]
Determined the cause 🤦🏻 We have an event service for propagating different events through our system (relies on Redis pub/sub) which was creating a new connection for every event... Thanks for your answers. It helped me to understand this was **not a BullMQ issue at all**.  Now to constrain our event service to re-use connections 😬
